### ISSUE TITLE
Make sure tt-metal can be built on Ubuntu 24.04

### ISSUE THREAD

**Issue by olofj** [tenstorrent/tt-metal#7921]
Ubuntu LTS 24.04 was just released, with a newer GCC and Python (3.12) that needs a few touchups for builds.

The newer GCC is slightly more picky on a few language constructs as well.

(Filing bug to tag changes with)

**Comment by tt-rkim** [comment]
@TT-billteng @jliangTT

Tagging for follow and triage. This may help lighten the load on our higher Python / Ubuntu versions efforts.

**Comment by olofj** [comment]
Separated into two pull requests (Python / C++), but I can split the C++ side into arbitrary number if it makes review and landing easier.

Both are up in draft state to run postcommit before marking as ready for review.

**Comment by blozano-tt** [comment]
Dockerfile was rewritten such that it should work for any ubuntu. Only issue might be gcc install from install_dependencies.sh could need massage.

**Comment by tt-rkim** [comment]
Could we just add extra handling for an even higher gcc version?

And do we have a related ticket to reduce dependency on libc++? Lower priority, but you said it's still good to do

**Comment by blozano-tt** [comment]
Dockerfile hit a problem with ubuntu 24.04 ... due to management of python dependencies without a veritual environment. So there will be work to do here.

**Comment by tt-rkim** [comment]
Should we retry this?

**Comment by blozano-tt** [comment]
Yeah, I thought I saw it working after we added venv into the docker image. However, need to double check.

If it works, we should add a 24.04 build into all-post-commit.

**Comment by tt-rkim** [comment]
We're now building this  and running smoke tests in merge gate.

I hope that fills everyone here with joy !!!
